File: backend/app/models/references.py
import uuid
from enum import Enum
from typing import Dict, List, Literal, Union

from pydantic import BaseModel, Field

from app.models.chapter import BasicChapter


class ChapterRefType(str, Enum):
    ABS = "abs"
    EMBEDDED = "embedded"
    AUDNEXUS = "audnexus"
    FILE_DATA = "file_data"
    JSON = "json"
    CSV = "csv"
    CUE = "cue"
    SNAPSHOT = "snapshot"


class TitleRefType(str, Enum):
    TEXT = "text"
    EPUB = "epub"
    MOBI = "mobi"
    CUSTOM = "custom"


class ReferenceBase(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    name: str
    short_name: str
    description: str
    metadata: Dict[str, str] = Field(default_factory=dict)


class ChapterReference(ReferenceBase):
    type: ChapterRefType
    chapters: List[BasicChapter]
    duration: float


class ReferenceValidationChapter(BasicChapter):
    headings: str = ""
    valid: bool = False


class ReferenceValidationResult(ReferenceBase):
    type: Union[ChapterRefType, Literal["intelligent_detection"]]
    chapters: List[ReferenceValidationChapter]
    duration: float


class TitleReference(ReferenceBase):
    type: TitleRefType
    titles: List[str]


Reference = Union[ChapterReference, TitleReference]

File: frontend/src/types/references.ts
import type { BasicChapter } from './chapter';

export type ChapterRefType = 'abs' | 'embedded' | 'audnexus' | 'file_data' | 'json' | 'csv' | 'cue' | 'snapshot';
export type ReferenceValidationType = ChapterRefType | 'intelligent_detection';
export type TitleRefType = 'text' | 'epub' | 'custom';

interface ReferenceBase {
  id: string;
  name: string;
  short_name: string;
  description: string;
  metadata: Record<string, string>;
}

export interface ChapterReference extends ReferenceBase {
  type: ChapterRefType;
  chapters: BasicChapter[];
  duration: number;
}

export interface ReferenceValidationChapter extends BasicChapter {
  headings: string;
  valid: boolean;
}

export interface ReferenceValidationResult extends ReferenceBase {
  type: ReferenceValidationType;
  chapters: ReferenceValidationChapter[];
  duration: number;
}

export interface TitleReference extends ReferenceBase {
  type: TitleRefType;
  titles: string[];
}

export type Reference = ChapterReference | TitleReference;
